data/transactions/logic/box.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+const (
+	boxCreate = iota
+	boxRead
+	boxWrite
+	boxDelete
+)
+
+func (cx *EvalContext) availableBox(name string, operation int, createSize uint64) error {
+	bt, ok := cx.available.boxes[boxRef{cx.appID, name}]
+	if !ok {
+		return fmt.Errorf("invalid Box reference %v", name)
+	}
+	switch operation {
+	case boxCreate:
+		bt.dirty = true
+		bt.size = createSize
+	case boxWrite:
+		bt.dirty = true
+	case boxDelete:
+		bt.size = 0
+	case boxRead:
+		/* nothing to do */
+	}
+	cx.available.boxes[boxRef{cx.appID, name}] = bt
+	return nil
+}
+
+func createBox(cx *EvalContext, name string, value string, appAddr basics.Address) (bool, error) {
+	// Enforce length rules. Currently these are the same as enforced by
+	// ledger. If these were ever to change in proto, we would need to isolate
+	// changes to different program versions. (so a v7 app could not see a
+	// bigger box than expected, for example)
+	if len(name) == 0 {
+		return false, fmt.Errorf("box names may not be zero length")
+	}
+	if len(name) > cx.Proto.MaxAppKeyLen {
+		return false, fmt.Errorf("name too long: length was %d, maximum is %d", len(name), cx.Proto.MaxAppKeyLen)
+	}
+	size := uint64(len(value))
+	if size > cx.Proto.MaxBoxSize {
+		return false, fmt.Errorf("box size too large: %d, maximum is %d", size, cx.Proto.MaxBoxSize)
+	}
+
+	err := cx.availableBox(name, boxCreate, size) // annotate size for write budget check
+	if err != nil {
+		return false, err
+	}
+
+	return cx.Ledger.NewBox(cx.appID, name, value, appAddr)
+}
+
+func opBoxCreate(cx *EvalContext) error {
+	last := len(cx.stack) - 1 // size
+	prev := last - 1          // name
+
+	name := string(cx.stack[prev].Bytes)
```

Comment:
Because our routines for emitting json use `[]byte` vs `string` as the trigger to decide whether to b64 encode the outputs.  Since we want to emit Transactions as json, making them `[]byte` simplifies things.  I had code to convert, but I found I needed to keep fighting and copying in many places to get things to work.